src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl

```diff
@@ -75,3 +78,144 @@ end
+
+
+    ## Compute volume scaled divergence of the barotropic transport and put into solver RHS
+    event = compute_volume_scaled_divergence!(free_surface, model)
+    wait(device(model.architecture), event)
+    
+    ## Include surface pressure term into RHS
+    RHS = free_surface.implicit_step_solver.solver.settings.RHS
+    RHS .= RHS/(model.free_surface.gravitational_acceleration*Δt)
+    ## fill_halo_regions!(RHS.data ,RHS.boundary_conditions, model.architecture, model.grid, model.clock, fields(model) )
+    η = free_surface.η
+    η = η_save
+    fill_halo_regions!(RHS   , η.boundary_conditions, model.architecture, model.grid)
+    fill_halo_regions!(η.data, η.boundary_conditions, model.architecture, model.grid)
+    ##  need to subtract Azᵃᵃᵃ(i, j, 1, grid)*η[i,j, 1]/(g*Δt^2)
+    event = add_previous_free_surface_contribution(free_surface, model, Δt )
+    wait(device(model.architecture), event)
+    fill_halo_regions!(RHS   , η.boundary_conditions, model.architecture, model.grid)
+    ## RHS .= RHS .+ free_surface.η.data/Δt
+
+    ## Then we can invoke solve_for_pressure! on the right type via calculate_pressure_correction!
+    x  = free_surface.implicit_step_solver.solver.settings.x
+    x .= η.data
+    fill_halo_regions!(x ,η.boundary_conditions, model.architecture, model.grid)
+    solve_poisson_equation!(free_surface.implicit_step_solver.solver, RHS, x)
+    ## exit()
+    fill_halo_regions!(x ,η.boundary_conditions, model.architecture, model.grid)
+    free_surface.η.data .= x
+
+    ## Once we have η we can update u* and v* with pressure gradient just as in pressure_correct_velocities!
+
+    ## The explicit form of this function defaults to returning an event, we do the same for now.
+    return event
+end
+
+
+using Oceananigans.Architectures: device
+using Oceananigans.Operators: ΔzC
+
+"""
+Compute the vertical integrated transport from the bottom to z=0 (i.e. linear free-surface)
+
+    `U^{*} = ∫ [(u^{*})] dz`
+    `V^{*} = ∫ [(v^{*})] dz`
+"""
+### Note - what we really want is RHS = divergence of the vertically integrated transport
+###        we can optimize this a bit later to do this all in one go to save using intermediate variables.
+function compute_vertically_integrated_transport!(free_surface, model)
+
+    event = launch!(model.architecture,
+                    model.grid,
+                    :xy,
+                    _compute_vertically_integrated_transport!,
+                    model.velocities,
+                    model.grid,
+                    free_surface.barotropic_transport,
+                    dependencies=Event(device(model.architecture)))
+
+    return event
+end
+
+@kernel function _compute_vertically_integrated_transport!(U, grid, barotropic_transport )
+    i, j = @index(Global, NTuple)
+    # U.w[i, j, 1] = 0 is enforced via halo regions.
+    barotropic_transport.u[i, j, 1] = 0.
+    barotropic_transport.v[i, j, 1] = 0.
```

Comment:
I think `0.` is `Float64` --- is this what we want? We may want to write just `0` here, which will be promoted to `eltype(barotropic_transport.u)`, etc.